greymatter-io/operator
# Sidecar injection reconciliation

[sc-7295]
[sc-22489]
[sc-22032]

Companion gitops-core PR: https://github.com/greymatter-io/gitops-core/pull/67

### To Test

1. Be on the `sidecar-injection-reconciliation` branch of the operator
2. Point the `pkg/cuemodule/core` submodule at your own branch off of the companion branch of gitops-core (though if that^ PR is already merged by then, a branch off of `main` will do).
3. Build the operator and push to a tag you control, (`reconcil_test` below), e.g.:
  ```
  scripts/build container && docker tag quay.io/greymatterio/operator:latest quay.io/greymatterio/operator:reconcil_test && docker push quay.io/greymatterio/operator:reconcil_test
  ```
4. Launch the operator (the following should do it):
  ```
  (
  cd pkg/cuemodule/core

  kubectl create namespace gm-operator

  kubectl create secret docker-registry gm-docker-secret \
    --docker-server=quay.io \
    --docker-username=$QUAY_USERNAME \
    --docker-password=$QUAY_PASSWORD \
    -n gm-operator

  kubectl create secret generic greymatter-sync-secret \
    --from-file=id_ed25519=$HOME/.ssh/id_ed25519 \
    -n gm-operator

  cue eval -c ./k8s/outputs --out text \
           -t operator_image=quay.io/greymatterio/operator:reconcil_test \
           -e operator_manifests_yaml | kubectl apply -f -
  )
  ```
5. Wait for the operator to deploy a mesh. This should happen more quickly than it did previously, because webhooks are no longer registered at all, and because of new optimizations that avoid reloading the CUE more often than necessary.
6. Once it's up, you can test reconciliation by deploying a workload into a watched namespace. Deploying the following (assumed to be named "deployment.yaml") with `kubectl apply -f deployment.yaml` should do it (**Potential Gotcha**: Notice that the annotations have moved back to the top-level `metadata` block of Deployments):
  ```
  # deployment.yaml
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: simple-server
    annotations:
      greymatter.io/inject-sidecar-to: "3000"
      greymatter.io/configure-sidecar: "true"
  spec:
    selector:
      matchLabels:
        app: simple-server
    template:
      metadata:
        labels:
          app: simple-server
      spec:
        containers:
          - name: server1
            image: python:3
            command: ["python"]
            args: ["-m", "http.server", "3000"]
  ```
7. This Deployment should receive a sidecar after a delay, and it should end up configured with Grey Matter configuration and end up with a card on the dashboard. The operator logs should describe the process of reconciliation. Multiple "reconciling workload" lines are expected. Also note that the original Deployment without a sidecar will appear briefly before reconciliation begins, and the associated pod will be Terminating for a while shortly afterwards.
8. Once the workload is up and running, delete it with `kubectl delete -f deployment.yaml`, and watch the operator logs to observe the removal process. It should end up removed from K8s and Grey Matter.
9. Deploy it again, and wait for simple-server to come back up.
10. Test GitOps updates by making a change to your target GitOps repository and pushing it. The operator should apply the update, and bump all Deployments and StatefulSets in watched namespaces to trigger an extra reconciliation, which causes your workload to update. (For now, it updates every time even if nothing has changed - this will be revisited once we have direction on desirable annotation function).
11. Delete the workload again as in step 8.
12. Tear down your mesh with the usual `cue eval ..... | kubectl delete -f -` command.
13. Set `spire` to `true` in your GitOps target, and push. Then redeploy with spire enabled, and redo steps 4-12 to confirm everything continues to function with spire enabled.

**Potential Gotcha**: If the operator is unavailable for reconciliation when you delete a workload it was managing, (for example, if you delete your entire mesh before deleting a managed workload) it may seem like you can't delete that workload. That's because the operator registers a "finalizer" on each of its managed workloads which pauses deletes until it can execute cleanup, and then it would remove the finalizer, but of course if the operator is unavailable, it will not be removed automatically. You can remove it manually by deleting the workload, and then _editing_ the workload to remove the `finalizers:` block, at which point the workload will continue to Termination:
```
metadata:
  ...
  finalizers:
    - greymatter.io/sidecar-finalizers
```

**Potential Gotcha**: Reconciliation sometimes triggers multiple deletes for the same action, which can cause "failed delete" log lines. These are not a problem and can be ignored if previous deletes have succeeded. The ultimate solution to this problem is to allow Control API to support idempotent deletes.
